### PR TITLE
Fixes to abilities + lots more abilities

### PR DESCRIPTION
**Pull request by ArtiOJ**
This is the current version on the server to my knowledge.

### CODE REVIEW

**Comment by ArtiOJ** [comment]
The merge conflict was resolved. Unless I am wrong, it should use the `==` operator instead of the `.equals()` function because I am trying to make sure that it is the same player, thus in the same spot in memory, if my understanding is correct, `.equals()` will only check if it is an instance of a player, which I do not want. I am checking to see if the living entity is a ring owner, not if the living entity is a Player, because obviously it is not. @RTClarkV

**Comment by RTClarkV** [comment]
Let him cook! – I'll have some time later today to take a look.

File: src/main/java/dev/corestone/lotrings/abilities/NecromanceAbility.java [hunk 1]
```diff
@@ -79,7 +79,10 @@ public void onHit(EntityDamageByEntityEvent event){
             Player player = ((Player) event.getDamager());
             LivingEntity damagedPlayer = (LivingEntity) event.getEntity();
 
-            for(LivingEntity e : getNecroEntities(player)){
+            if (damagedPlayer instanceof Player) {
```
Comment (RTClarkV) [inline src/main/java/dev/corestone/lotrings/abilities/NecromanceAbility.java:83]: I think you can just check the damager's UUID without checking if it's a player. Every entity has a UUID.

File: src/main/java/dev/corestone/lotrings/abilities/HeldPotionEffectAbility.java [hunk 2]
```diff
@@ -40,7 +40,4 @@ public void switchState(RingState ringState){
         if(ringState == RingState.HELD) Bukkit.getPlayer(ring.getOwner()).addPotionEffect(new PotionEffect(potionEffectType, 100000, power));
         if(ringState != RingState.HELD) Bukkit.getPlayer(ring.getOwner()).removePotionEffect(potionEffectType);
     }
-
```
Comment (RTClarkV) [inline src/main/java/dev/corestone/lotrings/abilities/HeldPotionEffectAbility.java:43]: ... fair.

File: src/main/java/dev/corestone/lotrings/abilities/FireballRainAbility.java [hunk 3]
```diff
@@ -97,12 +99,23 @@ public void shootFireball(Player player) {
...
         entityIDs.remove(event.getEntity().getUniqueId());
+        immune = true;
         if(explosionRadius > 0) event.getEntity().getWorld().createExplosion(event.getEntity(), explosionRadius, fire);
+        immune = false;
         if(event.getHitEntity() == null)return;
         if(event.getHitEntity() instanceof LivingEntity){
-            livingEntity livingEntity = (LivingEntity) event.getHitEntity();
+
+            LivingEntity livingEntity = (LivingEntity) event.getHitEntity();
             if (livingEntity instanceof Player && livingEntity == ring.getOwner()) return;
```
Comment (RTClarkV) [inline src/main/java/dev/corestone/lotrings/abilities/FireballRainAbility.java]: did you ever try to compile this? If so this line here should give an error. ring.getOwner() is a UUID and livingEntity is a LivingEntity.

I think this line of code only needs to see if the living entities' UUID equals the owner's UUID.

Comment (RTClarkV) [reply src/main/java/dev/corestone/lotrings/abilities/FireballRainAbility.java]: to make this work, we need to use livingEntity.getUUID()

**Pull request merged**